applications/GeoMechanicsApplication/custom_constitutive/coulomb_with_tension_cut_off_impl.cpp

```diff
@@ -112,34 +120,55 @@ bool CoulombWithTensionCutOffImpl::IsAdmissibleSigmaTau(const Vector& rTrialSigm
 }
 
 Vector CoulombWithTensionCutOffImpl::DoReturnMapping(const Vector& rTrialSigmaTau,
-                                                     CoulombYieldSurface::CoulombAveragingType AveragingType) const
+                                                     CoulombYieldSurface::CoulombAveragingType AveragingType)
 {
-    const auto apex = CalculateApex(mCoulombYieldSurface.GetFrictionAngleInRadians(),
-                                    mCoulombYieldSurface.GetCohesion());
-
-    if (IsStressAtTensionApexReturnZone(rTrialSigmaTau, mTensionCutOff.GetTensileStrength(), apex)) {
-        return ReturnStressAtTensionApexReturnZone(mTensionCutOff.GetTensileStrength());
-    }
-
-    const auto corner_point =
-        CalculateCornerPoint(mCoulombYieldSurface.GetFrictionAngleInRadians(),
-                             mCoulombYieldSurface.GetCohesion(), mTensionCutOff.GetTensileStrength());
-    if (IsStressAtTensionCutoffReturnZone(rTrialSigmaTau, mTensionCutOff.GetTensileStrength(), apex, corner_point)) {
-        return ReturnStressAtTensionCutoffReturnZone(
-            rTrialSigmaTau, mTensionCutOff.DerivativeOfFlowFunction(rTrialSigmaTau),
-            mTensionCutOff.GetTensileStrength());
-    }
-
-    if (IsStressAtCornerReturnZone(
+    Vector result = ZeroVector(2);
+
+    const double kappa_start = mCoulombYieldSurface.GetKappa();
+    double       error       = 1.0;
+    int          counter     = 0;
+    while (error > 1.0e-8) {
+        const auto apex = CalculateApex(mCoulombYieldSurface.GetFrictionAngleInRadians(),
+                                        mCoulombYieldSurface.GetCohesion());
```

Comment:
Since the location of the apex may now depend on $`\kappa`$, it may be wise to now make `CalculateApex` a `const` member function of class `CoulombYieldSurface`. In that case, we no longer need to pass the friction angle $`\phi`$ and cohesion $`c`$, since the class can retrieve those values by calling its members `GetFrictionAngleInRadians` and `GetCohesion` from the implementation of member `CalculateApex`. The code here would then look as follows:
```suggestion
        const auto apex = mCoulombYieldSurface.CalculateApex();
```